pkg/metadata.go

```diff
@@ -0,0 +1,131 @@
+}
+
+func GetFieldValueByNameError(target string) error {
+	return errors.Error("GetFieldValueByName('%v') failed", target)
+}
+
+func (from NlxMetadata) GetFieldValueByName(target string) (names []string, err error) {
+	mType := reflect.TypeOf(NlxMetadata{})
+
+	var name string
+	for i := 0; i < mType.NumField(); i++ {
+		field := mType.Field(i)
+		if field.Tag.Get("json") == target {
+			name = field.Name
+			break
+		}
+	}
+
+	if name == "" {
+		err = GetFieldValueByNameError(target)
+		return
+	}
+
+	value := reflect.ValueOf(from)
+	field := value.FieldByName(name)
+	if field.IsValid() {
+		names = field.Interface().([]string)
+		return
+	}
+
+	logging.Get().Tracef("Somehow able to find field name %v but not its value as []string in the metadata", name)
+	err = GetFieldValueByNameError(target)
+
+	return
+}
+
+func (m NlxMetadata) FilterItem(item string) (keep bool) {
+	cleanRelativeFilePath := util.FromWindowsPath(item)
+	directory := filepath.Dir(cleanRelativeFilePath)
+	baseName := filepath.Base(cleanRelativeFilePath)
+
+	// Find the lowest level folder
+	dirSplit := strings.Split(directory, string(filepath.Separator))
+	metadataType, subFolders := dirSplit[0], dirSplit[1:]
+	filePathArray := append(subFolders, baseName)
+	filePath := strings.Join(filePathArray, string(filepath.Separator))
+
+	validMetadataNames, err := m.GetFieldValueByName(metadataType)
+	if validMetadataNames == nil || len(validMetadataNames) == 0 {
+		logging.Get().Trace("No valid names for this directory.")
+		return
+	}
+
+	if err != nil {
+		logging.Get().Errorf("Metadata Filter Error: %v", err)
+		return
+	}
+
+	if util.StringSliceContainsAnyKey(validMetadataNames, []string{
+		// Most common case --- check for our metadata with .json stripped
+		strings.TrimSuffix(filePath, ".json"),
+		// See if our filePath is in the valid metadata, if so, we're done
+		filePath,
+	}) {
+		keep = true
+		return
+	}
+
+	// Check for children of a component pack
+	if metadataType == "componentpacks" {
+		filePathParts := strings.Split(filePath, string(filepath.Separator))
+		if len(filePathParts) == 2 && util.StringSliceContainsKey(validMetadataNames, filePathParts[0]) {
+			logging.Get().Tracef("Keeping componentpack metadata file: %v", filePath)
+			keep = true
+			return
+		}
+	}
+
+	if util.StringSliceContainsAnyKey(validMetadataNames, []string{
+		// Check for our metadata with .xml stripped
+		strings.TrimSuffix(filePath, ".xml"),
+		// Check for our metadata with .skuid.json stripped
+		strings.TrimSuffix(filePath, ".skuid.json"),
+		// Check for theme inline css
+		strings.TrimSuffix(filePath, ".inline.css"),
+	}) {
+		logging.Get().Tracef("Keeping metadata file: %v", filePath)
+		keep = true
+		return
+	}
+
+	return
+}
+
+// GetMetadataTypeDirNames returns the directory names for a type
+func GetMetadataTypeDirNames() (types []string) {
+	metadataType := reflect.TypeOf(NlxMetadata{})
+
+	for i := 0; i < metadataType.NumField(); i++ {
```

Comment:
Since here it's being treated with reflection, I suggest to store the NumField() in a new variable and reuse it.

Like:

`for i, nF := 0, metadataType.NumField(); i < nF; ...`